File: src/004/p004.rs
fn is_palindrome(n: i32) -> bool {
    let reversed = n.to_string().chars().rev().collect::<String>();
    reversed == n.to_string()
}

fn solve() -> i32 {
    let mut largest = 0;
    for x in 100..999 {
        for y in 100..999 {
            let product = x * y;
            if product > largest && is_palindrome(product) {
                largest = product;
            }
        }
    }
    largest
}

fn main() {
    let result = solve();
    println!("{}", result);
}

File: src/004/p004.lua
function is_palindrome(n)
  local str = tostring(n)
  return str == string.reverse(str)
end

function solve()
  local largest = 0
  for a=0, 999 do
    for b=0, 999 do
      local product = a * b
      if product > largest and is_palindrome(product) then
        largest = product
      end
    end
  end
  return largest
end

local result = solve()
print(result)
